Add BrayCurtis

kescobo opened the pull request:
Addresses #75 

I modeled this off of the jaccard code, since they're relatively similar. 

This still needs tests, but I could use some guidance as to whether the way I've done this makes sense, and what tests I should add.

KristofferC commented:
TODO: Test correctness on some points including some degenerate cases,  and adding the distance to the table in the README. Should be good after that,

kescobo commented:
@KristofferC Ahh, sorry I missed the table - fixed. I presume someone else will run the benchmarks so it's consistant with the others? 

Would you please clarify what you mean by "degenerate cases?" I added a couple of things that I think are along those lines... does that satisfy what you meant, or was there something else?

nalimilan commented:
@KristofferC Bump.

kescobo commented:
Thanks for the bump @nalimilan :-)

@KristofferC I'm hoping to do an initial release of a microbiome package that's somewhat dependent on this PR. Any chance you could review the test cases I added are sufficient?

alyst commented on README.md:
```diff
@@ -128,37 +129,38 @@ This type system has practical significance. For example, when computing pairwis
+|  MeanSqDeviation          |  `msd(x, y)`               | `mean(abs2.(x - y))` |
+|  RMSDeviation             |  `rmsd(x, y)`              | `sqrt(msd(x, y))` |
+|  NormRMSDeviation         |  `nrmsd(x, y)`             | `rmsd(x, y) / (maximum(x) - minimum(x))` |
+| Bray-Curtis Dissimliarity |  `braycurtis(x, y)`        | `sum(abs(x - y)) / sum(abs(x+y))`  |
```
IIUC, it should be just BrayCutris, since in this column you provide the type names to be used from e.g. `pairwise()` method.
So increasing the column width would be also unnecessary.

kescobo replied:
@alyst Sorry for delayed response (holiday). I'll take a look at this today.

kescobo replied:
Fixed.

alyst commented on test/test_dists.jl:
```diff
@@ -169,6 +169,13 @@ end
+        @test braycurtis([1,0,3],[0,1,0]) == 1.0
+        @test braycurtis(rand(10), zeros(10)) == 1.0
+        @test braycurtis([1,0],[1,0]) == 0.0
+
```
I guess `Bool` case has to be tested as well.
Also, there's `T`, `_x`, `_y` loop above, I think it makes sense to add `braycurtis()` tests there as well (that should test `Float64` case).

kescobo replied:
I added `braycurtis()` to the loop, as suggested, but I don't think Bray Curtis is defined for the boolean case, or at least it's not really an appropriate dissimilarity measure for presence/absence type situations.

alyst replied:
You have defined `eval_start()` for a `Bool` case, so I supposed it have to be tested.
E.g `Euclidean` or `Cityblock` also support boolean vectors.

alyst commented on src/metrics.jl:
```diff
@@ -356,6 +357,27 @@ end
+    a, b
+end
+@inline function eval_end(::BrayCurtis, a)
+    @inbounds v = a[1] / a[2]
```
Should `a[1] == 0 && a[2] == 0` case give NaN or 0.0?

kescobo replied:
It should give NaN (or throw an error) - rows where both counts are zero should not be included in the calculation.

kescobo commented:
Thanks for the review @alyst - I've made the fixes you suggested, except for adding a test for the boolean case, as I don't think that would be an appropriate use of this distance metric. If you think it should be supported anyway, I'd be happy to add it in.

kescobo commented:
@alyst OK - I added a test for the boolean case and rebased. Tests passed locally, so I'm guessing travis will give the green light. Thanks again for taking this on!

alyst commented:
Yes, I guess it would be better to update the benchmarks in README.md using the same machine.
But meanwhile you can add BrayCurtis to the benchmark scripts and post the results here.

kescobo commented:
I've updated the benchmark files, but was unable to complete the run. Complete output [is here](https://gist.github.com/kescobo/c2de44e7bee52382c05f94c86c6bfe30), but the relevant error is:

```
ERROR: LoadError: ArgumentError: JLD serialization is no longer supported. Benchmarks should now be saved in
JSON format using `save("params".json, args...)` and loaded from JSON using
using `load("params".json, args...)`. You will need to convert existing
saved benchmarks to JSON in order to use them with this version of BenchmarkTools.
Stacktrace:
 [1] badext(::String) at /Users/kev/.julia/v0.6/BenchmarkTools/src/serialization.jl:61
 [2] save(::String, ::String, ::Vararg{Any,N} where N) at /Users/kev/.julia/v0.6/BenchmarkTools/src/serialization.jl:65
 [3] include_from_node1(::String) at ./loading.jl:576
 [4] include(::String) at ./sysimg.jl:14
 [5] process_options(::Base.JLOptions) at ./client.jl:305
 [6] _start() at ./client.jl:371
while loading /Users/kev/.julia/v0.6/Distances/benchmark/print_table.jl, in expression starting on line 43
```

kescobo commented:
@alyst Seems like the above error is a problem with the benchmarking script rather than with the PR, right? I think this is ready to go.

alyst commented:
@kescobo Yes, I'm just waiting to collect feedback :) Otherwise I'll merge it around 22:00 CET
